// map-factory/dist/test/array-value-arrays-test.js
"use strict";

var _lab = require("lab");

var Lab = _interopRequireWildcard(_lab);

var _notationSuite = require("./suites/notation-suite");

var _notationSuite2 = _interopRequireDefault(_notationSuite);

var _manyMappingsSuite = require("./suites/many-mappings-suite");

var _manyMappingsSuite2 = _interopRequireDefault(_manyMappingsSuite);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var lab = exports.lab = Lab.script();

_notationSuite2.default.run(lab, {
  LABELS: ["arrays", "value arrays", "single mapping array specifying index"],
  GET_ITEM: "item",
  SET_ITEM: "[0]",
  SOURCE: {
    item: 123
  },
  EXPECTED: [123],
  NO_SOURCE_EXPECTED: [],
  MODIFY_VALUE: "modified",
  MODIFIED_EXPECTED: ["modified"]
});

_notationSuite2.default.run(lab, {
  LABELS: ["arrays", "value arrays", "single mapping array"],
  GET_ITEM: "item",
  SET_ITEM: "[]",
  SOURCE: {
    item: 123
  },
  EXPECTED: [123],
  NO_SOURCE_EXPECTED: [],
  MODIFY_VALUE: "modified",
  MODIFIED_EXPECTED: ["modified"]
});

// size is missing from source
var labels = ["arrays", "value arrays", "mix of available and missing data"];
var mappings = [{ from: "group.colours.[]", to: "item.signals" }, { from: "group.name", to: "item.name" }];
var source = {
  "group": {
    "colours": ["red", "yellow", "green"]
  }
};

var expected = {
  "item": {
    "signals": ["red", "yellow", "green"]
  }
};

// manyMappings.run(lab, {
//   LABELS: labels,
//   MAPPINGS: mappings,
//   SOURCE: source,
//   EXPECTED: expected,
//   EXPERIMENTAL: true
// });

_manyMappingsSuite2.default.run(lab, {
  LABELS: labels,
  MAPPINGS: mappings,
  SOURCE: source,
  EXPECTED: expected,
  EXPERIMENTAL: false
});
